File: 백준/Gold/1943. 동전 분배/동전 분배.py
# DP

for _ in range(3):
    N = int(input())
    coins = []
    total_coin = 0
    for _ in range(N):
        coin, cnt = map(int, input().split())
        total_coin += coin * cnt
        coins.append((coin, cnt))
    # 홀수는 반으로 나눌 수 없음 -> 확인 안해봐도 됨
    if total_coin % 2 == 1:
        print(0)
        continue
    total_coin = total_coin // 2

    # dp[i] : 동전들로 i원 만들수 있는지 여부
    dp = [False] * (total_coin + 1)
    dp[0] = True

    answer = 0

    for coin, cnt in coins:
        # 뒤에서부터 탐색하면서 지불 가능한 액수 갱신
        for n in range(total_coin, coin-1, -1):
            if dp[n-coin]:
                for j in range(cnt):
                    if n + coin * j <= total_coin:
                        dp[n + coin * j] = True
                    else:
                        break
        if dp[-1]:
            answer = 1
            break
    print(answer)

File: 백준/Gold/1943. 동전 분배/동전 분배.java
import java.util.*;
import java.io.*;

class Coin {
    int value, quantity;
    Coin(int value, int quantity) {
        this.value = value;
        this.quantity = quantity;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        for (int t = 0; t < 3; t++) {
//            boolean[] dp = new boolean[100001];
//            dp[0] = true;
            int sum = 0;
            int N = Integer.parseInt(br.readLine());
            Coin[] coins = new Coin[N+1];
            for (int i = 0; i < N; i++) {
                StringTokenizer st = new StringTokenizer(br.readLine(), " ");
                int coin = Integer.parseInt(st.nextToken());
                int cnt = Integer.parseInt(st.nextToken());
                coins[i+1] = new Coin(coin, cnt);
                sum += coin * cnt;
//                for (int k = 1; k <= cnt; k++) {
//                    dp[k * coin] = true;
//                }
            }
            if (sum % 2 == 1) System.out.println(0);
            else {
                boolean[] dp = new boolean[100001];
                dp[0] = true;
                for (int i = 1; i <= N; i++) {
                    int coin = coins[i].value;
                    int cnt = coins[i].quantity;

                    for (int j = sum/2; j >= coin; j--) {
                        if (dp[j - coin]) {
                            for (int p = 0; p < cnt; p++) {
                                if (j + coin * p <= sum/2){
                                    dp[j + coin * p] = true;
                                }
                                else {
                                    break;
                                }
                            }
                        }
                    }
                }
                if (dp[sum/2]) {
                    System.out.println(1);
                } else {
                    System.out.println(0);
                }
            }
        }
    }
}
